CP-49135: speed up UUID generation

edwintorok opened the pull request:
2 optimizations here:
* open /dev/urandom just once, speeds up session uuid/ref creation by ~3.8x
* use a PRNG for non-secret UUIDs. We cannot use these for sessions (those UUIDs/opaquerefs are effectively the authentication tokens), or Pool secrets, but we can use it for everything else.

The latter is not yet enabled by default, needs more testing/auditing on whether we have any other secret opaquerefs/UUIDs in the codebase, but the 1st one is enabled by default (we still use `/dev/urandom` for generating the UUIDs, we just don't keep closing and reopening it).

edwintorok commented:
Back to draft, this seems to reveal a bug somewhere else in the codebase (someone closed the wrong file descriptor and ended up closing the urandom one...)

edwintorok commented:
> Back to draft, this seems to reveal a bug somewhere else in the codebase (someone closed the wrong file descriptor and ended up closing the urandom one...)

Ah crowbar runs `at_exit` so I can't use `at_exit` to close this, because that'll then run *before* the Crowbar tests, making them all fail (fixed).

And forkexecd isn't happy about the extra FD either.

edwintorok commented on ocaml/libs/uuid/uuidx.mli:
```diff
@@ -33,7 +33,14 @@ val null : 'a t
 val make : unit -> 'a t
 (** Create a fresh UUID *)
 
+val make_uuid_fast : unit -> 'a t
```
Hmm I'd need a way to negate `'session', in the polymorphic variant, but I don't think OCaml's type system supports constraints like that. I can define upper or lower bounds that would include a conflicting `'session of 'somethingbad`, but that fails when an exact Ref.t is required.

It is quite easy to declare that make_secret returns a 'session Ref.t, but I'll keep thinking on whether there is a way to say that 'make' returns an 'a that can be anything exception 'session.

edwintorok replied:
I found a way, and it has also found a bug: the local session must also use the urandom method for generating the secret.

(had to tweak a few other things in mlis and testcases too).

There aren't that many "classes" in XAPI, so I've enumerated them all (by hand for now) in Uuidx.mli, and ensured that session cannot be present there (by adding session with an incompatible type, so that when unified it prints a nice error message):
```
    ```
    File "ocaml/tests/test_client.ml", line 18, characters 47-57:
    18 |       Test_common.make_session ~__context ~ref:session_id
                                                        ^^^^^^^^^^
    Error: This expression has type [< Uuidx.not_session ] Ref.t
           but an expression was expected of type [ `session ] Ref.t
           Type [< Uuidx.not_session ] = [< `session of [ `use_make_uuid_urnd ] ]
           is not compatible with type [ `session ]
           Types for tag `session are incompatible
    ```
```

edwintorok replied:
@mg12 you were also asking whether we can use the typesystem for tracking secrets, see above.

last-genius replied:
Does this mean that adding a new class to XAPI would require adding it to these two places too? Do we not produce some list of classes automatically? There's already a bit of boilerplate that needs to be written with the new class, and I don't think an error will show up here if one forgets to add the class.

edwintorok replied:
I've changed a bit how the parameter is handled, now there is a type constraint on `'a`, and it must be one of the known classes. If we add a new class then the generated db_actions.ml will ensure we get a type error, because it'll try to use a type parameter that is not in one of the known ones.

Eventually it'd be nice if we could autogenerate this list of classes, although we add classes rare enough.

psafont commented:
Thanks, It's interesting to clearly see the types weren't used properly before, and now they are being tracked as they should

edwintorok commented:
I've split the type safety changes into their own commits, and also introduced a 'Ref.of_secret_string' to more clearly separate the two.

Of course the type system can't tell us if we use secret UUIDs anywhere other than Session, it is up to us to tell it the list of classes that have secret uuids.
Also if you immediately convert a Uuidx.t to a string you wouldn't get an error that you generated it wrongly (but usually Ref and UUID generation are close together, as in the session code).

edwintorok commented on ocaml/idl/ocaml_backend/gen_api.ml:
```diff
@@ -400,15 +400,7 @@ let gen_client_types highapi =
-            x);"
-         ; "end"
-         ]
+       ; ["module Ref = Ref"]
```
I think I got a build failure without it, we could probably find what generates Foo.Ref and make it generate just Ref instead.
